domain/service/employees.go

```diff
@@ -0,0 +1,187 @@
+			continue
+		}
+		if err := stream.Send(&proto.EmployeeResponse{
+			Employee: entities.EmployeeFromStorageToProto(saved),
+		}); err != nil {
+			return err
+		}
+		savedCount++
+	}
+	if errorMessage != "" {
+		return errors.New(errorMessage)
+	}
+	log.Printf("successfully saved %d employees", savedCount)
+	return nil
+}
+
+func (e *employeeService) AddAttachment(stream proto.EmployeeService_AddAttachmentServer) error {
+	md, ok := metadata.FromIncomingContext(stream.Context())
+	var userID string
+	var fileName string
+	if ok {
+		userID = md.Get("userID")[0]
+		fileName = md.Get("filename")[0]
+		emp, err := e.repository.GetEmployee(userID)
+		if err != nil {
+			return fmt.Errorf("failed to get employeee with id \"%s\":%w", userID, err)
+		}
+		log.Printf("receiving photo for user %s (%s %s)\n", emp.ID, emp.FirstName, emp.LastName)
+	}
+	imgData := []byte{}
+	for {
+		data, err := stream.Recv()
+		if err == io.EOF {
+			log.Printf("file received with length: %d\n", len(imgData))
+			document, err := e.documents.SaveDocument(userID, fileName, imgData)
+			document.UserID = userID
+			if err != nil {
+				return fmt.Errorf("failed to save document: %w", err)
+			}
+			if err := e.repository.AddDocument(userID, document.ID); err != nil {
+				return fmt.Errorf("failed to save document for employee \"%s\": %w", userID, err)
+			}
+			return stream.SendAndClose(entities.DocumentFromStorageToProto(document))
+		}
+		if err != nil {
+			return fmt.Errorf("failed to read file data from stream: %w", err)
+		}
+		log.Printf("received %d bytes\n", len(data.Data))
+		imgData = append(imgData, data.Data...)
+	}
+}
+
+func (e *employeeService) AttachmentByID(ctx context.Context, req *proto.ByIDRequest) (*proto.Attachment, error) {
+	doc, err := e.documents.GetDocument(req.ID)
+	if err != nil {
+		return nil, err
+	}
+	return entities.DocumentFromStorageToProto(doc), nil
+}
+
+func (e *employeeService) DeleteEmployee(ctx context.Context, req *proto.ByIDRequest) (*proto.EmployeeResponse, error) {
+	employee, err := e.repository.DeleteEmployee(req.ID)
+	if err != nil {
+		return nil, err
+	}
+	if err := e.repository.DeleteVacations(req.ID); err != nil {
+		return nil, fmt.Errorf("failed to remove vacations: %w", err)
+	}
+	return &proto.EmployeeResponse{
+		Employee: entities.EmployeeFromStorageToProto(employee),
+	}, nil
+}
+
+func (e *employeeService) UpdateEmployee(ctx context.Context, req *proto.EmployeeUpdateRequest) (*proto.EmployeeResponse, error) {
+	employee, err := e.repository.UpdateEmployee(req.ID, entities.EmployeeFromProtoToStorage(req.Updates.Employee))
+	if err != nil {
+		return nil, err
+	}
+	return &proto.EmployeeResponse{
+		Employee: entities.EmployeeFromStorageToProto(employee),
+	}, nil
+}
+
+func (e *employeeService) AddVacation(ctx context.Context, req *proto.VacationRequest) (*proto.Vacation, error) {
+	start := time.Unix(req.StartDate, 0)
+	hours24 := time.Now().Add(24 * time.Hour)
+
+	if !start.After(hours24) {
```

Comment:
Might be worth thinking about converting to an internal entity structure to use within the application. Then you can add validation functions on that structure (see http://go-playground/validator for example). This would make the code more testable. Currently this function is doing several things, extracting data from proto request, validation, writing to storage, and conversion back to proto, which could make it hard to test.

Reply:
Link is not valid, but totally agree that it should be separated to layers.